rhizome/lib/storage_key_tool.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require "fileutils"
+require_relative "common"
+require_relative "vm_path"
+require_relative "../lib/storage_key_encryption"
+
+class StorageKeyTool
+  def initialize(vm_name, disk_index)
+    vp = VmPath.new(vm_name)
+    @key_file = vp.data_encryption_key(disk_index)
+    @new_key_file = "#{@key_file}.new"
+  end
+
+  def reencrypt_key_file(old_key, new_key)
+    sek_old = StorageKeyEncryption.new(old_key)
+    sek_new = StorageKeyEncryption.new(new_key)
+
+    data_encryption_key = sek_old.read_encrypted_dek(@key_file)
+
+    sek_new.write_encrypted_dek(@new_key_file, data_encryption_key)
+  end
+
+  def test_keys(old_key, new_key)
+    sek_old = StorageKeyEncryption.new(old_key)
+    sek_new = StorageKeyEncryption.new(new_key)
+
+    old_dek = sek_old.read_encrypted_dek(@key_file)
+    new_dek = sek_new.read_encrypted_dek(@new_key_file)
+
+    if old_dek[:cipher] != new_dek[:cipher]
+      raise "ciphers don't match"
+    end
+
+    if old_dek[:key] != new_dek[:key]
+      raise "keys don't match"
+    end
+
+    if old_dek[:key2] != new_dek[:key2]
+      raise "second keys don't match"
+    end
+  end
+
+  def retire_old_key
+    FileUtils.mv @new_key_file, @key_file
+    r "sync #{@key_file.shellescape}"
```

Comment:
You could do this, but it would be more parsimonious  to fysnc from ruby directly.

Also, to make renames persistent, you have to fsync the *directory* file descriptor. You should fsync the key file itself having run `write()`  of the encrypted material, but the promotion via `rename` requires fsync on something entirely different.

Examples:

```ruby
> File.open("/tmp/foo", "w") { _1.write("my data"); _1.flush; _1.fsync }
=> 0
> File.rename("/tmp/foo", "/tmp/bar") 
=> 0
# if you crash here, the rename may or may not persist, even if the `rename()` syscall did return 0.
[8] clover-development(main)> File.open("/tmp") { _1.fsync }
=> 0
```

I use the block form to ensure the file descriptor does not leak.

Note if at any point you receive non-zero, we should probably page, as The End is near, e.g. https://wiki.postgresql.org/wiki/Fsync_Errors. Since we don't have an obvious way to know what to do about that yet, at the very least, the program should not blithely continue with a non-zero exit code, perhaps crashing and not progressing makes sense. We'd have to think hard about failures at each possible sync and not wedging the system irrecoverably though.

If you want a fun bit of ancient history (so, it was contemporary when I was young), Linux's ext4 delayed allocation behavior caused a lot of consternation at the time because of the de-ordering between data flush and directory entry flush. https://lwn.net/Articles/322823/

Reply:
Done. Couple of notes:
* Ruby's fsync does fflush then fsync, so manual flush is redundant.
* Ruby's fsync doesn't return non-zero, but throws exception. I created helper functions to convert it to an "FsyncFail" exception so we can page if we catch it.

Reply:
I wasn't sure if fsync was kind enough to also handle buffered I/O...how convenient.